### DISCUSSION TITLE
Help on implementing ReserveRequest

### DISCUSSION THREAD

**Discussion by jxtaaa** [villekr/ocpp-asgi#24] [Q&A] [answered]
Hello there,

my name is José Luis Bueno, I’m currently studying Software Engineering in Málaga, Spain. I’m developing a webapp that manages bookings and some sort of things related to electric chargers and I found your repo.

Firstly, I want to thank you for your great work you did on this repository, it’s excellent and it’s helping me a lot and simplifying my webapp development.

This webapp is for a subject based on teaching protocols so I needed to use OCPP and I found your implementation of ASGI based on OCPP. I got a (surely very dumb) question but I’ve been stuck trying different methods and I can only get errors from my app.

What I want to do is to send a “ReserveNow” transaction. I got the following code on client_api.py:

![image](https://user-images.githubusercontent.com/74137433/229190076-ec3d7a4d-ec1a-48aa-84c3-a1043a125c11.png)

So it calls the following async method on my charging station (note: i’m using ocpp 201):

![image](https://user-images.githubusercontent.com/74137433/229190122-8e3d0e36-1163-4896-b910-5a123fbfe3eb.png)

In this method I call the central station to ask for the response of the transaction, here you have the function on provisioning_router.py:

![image](https://user-images.githubusercontent.com/74137433/229190135-32ff9fb3-6f14-4faa-82c3-19a266afd213.png)

Once I execute the POST call in the client api I’m getting three errors on the central system console:
1. jsonschema.exceptions.ValidationError: Additional properties are not allowed ('status' was unexpected). I don’t really understand why I’m getting this error, I gotta attach the result of the reservation and it’s gotta be made by the central station.
2. ocpp.exceptions.FormatViolationError: FormatViolationError: Payload for Action is syntactically incorrect or structure for Action, {'cause': "Additional properties are not allowed ('status' was unexpected)”}. The same as above.
3. [on_receive - 201] Failure when processing message on_receive: e=TypeError('asdict() should be called on dataclass instances’). I don’t either understand this error, I’m  providing the ReserveNowPayload dataclass, so why cant .asdict() be called?

And if I check the charging station console:
[_get_specific_response - 315] Ignoring response with unknown unique id: <CallError - unique_id=139c4075-372b-42e2-b36e-ac18bdb33f48, error_code=FormatViolation, error_description=Payload for Action is syntactically incorrect or structure for Action, error_details={'cause': "Additional properties are not allowed ('status' was unexpected)”}> It is the same as above, it’s spreading the exception.
Then I reveive a timeout exception.

Finally, on the client API console I get a 504 timeout, obviously.

Hope you can provide me some help to solve this problem. I’m only asked to implement a few methods but following thoroughly the OCPP documentation.

**Comment by villekr** [comment]
Hi,

Nice to hear that you've found ocpp-asgi library useful.

I'd say your problem is in a way how you're handling ReserveNow event. When your charging station receives a request from csms (ReserveNow) then within on-handler the only thing you can send back to csms is the response for that request. Here you're making a request back to csms with await self.call(request). Furthermore, ReserveNow request can only be made by csms to charging station.

If you want to perform any additional work after responding back to csms request you can do that in @after handler.

br,
VK
________________________________
From: José Luis Bueno ***@***.***>
Sent: Friday, March 31, 2023 20:33
To: villekr/ocpp-asgi ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [villekr/ocpp-asgi] Help on implementing AuthorizeRequest (Discussion #24)


Hello there,

my name is José Luis Bueno, I’m currently studying Software Engineering in Málaga, Spain. I’m developing a webapp that manages bookings and some sort of things related to electric chargers and I found your repo.

Firstly, I want to thank you for your great work you did on this repository, it’s excellent and it’s helping me a lot and simplifying my webapp development.

This webapp is for a subject based on teaching protocols so I needed to use OCPP and I found your implementation of ASGI based on OCPP. I got a (surely very dumb) question but I’ve been stuck trying different methods and I can only get errors from my app.

What I want to do is to send a “ReserveNow” transaction. I got the following code on client_api.py:

[image]<https://user-images.githubusercontent.com/74137433/229190076-ec3d7a4d-ec1a-48aa-84c3-a1043a125c11.png>

So it calls the following async method on my charging station (note: i’m using ocpp 201):

[image]<https://user-images.githubusercontent.com/74137433/229190122-8e3d0e36-1163-4896-b910-5a123fbfe3eb.png>

In this method I call the central station to ask for the response of the transaction, here you have the function on provisioning_router.py:

[image]<https://user-images.githubusercontent.com/74137433/229190135-32ff9fb3-6f14-4faa-82c3-19a266afd213.png>

Once I execute the POST call in the client api I’m getting three errors on the central system console:

  1.  jsonschema.exceptions.ValidationError: Additional properties are not allowed ('status' was unexpected). I don’t really understand why I’m getting this error, I gotta attach the result of the reservation and it’s gotta be made by the central station.
  2.  ocpp.exceptions.FormatViolationError: FormatViolationError: Payload for Action is syntactically incorrect or structure for Action, {'cause': "Additional properties are not allowed ('status' was unexpected)”}. The same as above.
  3.  [on_receive - 201] Failure when processing message on_receive: e=TypeError('asdict() should be called on dataclass instances’). I don’t either understand this error, I’m providing the ReserveNowPayload dataclass, so why cant .asdict() be called?

And if I check the charging station console:
[_get_specific_response - 315] Ignoring response with unknown unique id: <CallError - unique_id=139c4075-372b-42e2-b36e-ac18bdb33f48, error_code=FormatViolation, error_description=Payload for Action is syntactically incorrect or structure for Action, error_details={'cause': "Additional properties are not allowed ('status' was unexpected)”}> It is the same as above, it’s spreading the exception.
Then I reveive a timeout exception.

Finally, on the client API console I get a 504 timeout, obviously.

Hope you can provide me some help to solve this problem. I’m only asked to implement a few methods but following thoroughly the OCPP documentation.

**Reply by jxtaaa** [reply]
Thank you for your answer. Here I got a few questions: provisioning_router.py is acting as the CSMS, right? What I really want to do is to follow this diagram (OCPP docs):

<img width="1286" alt="image" src="https://user-images.githubusercontent.com/74137433/229534679-612bd0f1-c0a9-4155-a355-84a08af54410.png">

So what I created is a POST endpoint on the client api as it follows:

```python
@app.post(
    "/reserve_now/{charging_station_id}",
    response_model=call_result.ReserveNowPayload,
    status_code=status.HTTP_200_OK,
    tags=["ocpp version 2.0.1"],
)
async def reserve_now(charging_station_id: int, expiry_date_time: datetime, token: str):
    payload = call.ReserveNowPayload(
        id = 1,
        expiry_date_time = expiry_date_time.isoformat(),
        id_token = {
            "idToken": token,
            "type": "Local"
        },
    )
    try:
        response: call_result.ReserveNowPayload = await post_to_connection(
            charging_station_id=charging_station_id,
            payload=payload,
            ocpp_version=OCPPVersion.v2_0_1,
        )
        return response
    except TimeoutError:
        raise HTTPException(
            status_code=status.HTTP_504_GATEWAY_TIMEOUT,
            detail="No response from Charging station",
        )
```

(the rest of the code is the same as I posted last time, I haven't changed anything yet)

When I call `post_to_connection` what I'm really doing is calling the CSMS or the charging station method itself? Can I choose to call either CSMS or charging station from client API? What order should I follow to really stick to he OCPP docs? 

Thank you in advance, your help is being really meaningful to my work.

**Reply by villekr** [reply]
From the client API request handler you initiate communication towards charging station (ReserveNow request). Then when charging station responds (ReserveNow response) that message is directly routed where response is being waited on (await post_to_connection call). This `post_to_connection` does three things for you:

1. Post a request to callback api
2. Wait for response from the charging station
3. Construct response payload as appropriate dataclass instance

Whether you run your "architecture" using standalone or serverless the details how this happens is a little different. But in client API you don't need to care about that.

So basically:

- From charging station's `on_reserve_now` function remove the `... await self.call...` call. You just respond with ReserveNow response and with proper status value whether your charging station is able to accept reservation or not.
- Remove router `on_reserve_now` function. There's no ReserveNow request in OCPP that is sent from charging station to csms. It's other way around.

**Reply by jxtaaa** [reply]
Thank you again for replying, that's just what I was thinking to do with the code. Now, I'd like to ask you if there's a way to authorize users (not exactly if there's a way, just if there's a way that fits my case). On a first approach I had thought about securizing each client API endpoint as everyone could do using FastAPI, but I'm required to follow OCPP docs (again... haha) so I need to authorize my users using an accessToken I've previously given to them (context: I'm using OCPP client API in a webapp with previous authentication & authorization, but I need to make authorized OCPP calls to the client API). I'm using the AuthorizaRequest use case between charging station and CSMS:

`client_api.py` endpoint:
```python
@app.post(
    "/authorize/{charging_station_id}",
    response_model=call_result.AuthorizePayload,
    status_code=status.HTTP_200_OK,
    tags=["ocpp version 2.0.1"],
)
async def authorize(charging_station_id: str, token: str):
    payload = call.AuthorizePayload(id_token={
        "idToken": token,
        "type": "Central"
    })
    try:
        response: call_result.AuthorizePayload = await post_to_connection(
            charging_station_id=charging_station_id,
            payload=payload,
            ocpp_version=OCPPVersion.v2_0_1,
        )
        return response
    except TimeoutError:
        raise HTTPException(
            status_code=status.HTTP_504_GATEWAY_TIMEOUT,
            detail="No response from Charging station",
        )
```

`charging_station.py` Authorize handler
```python
@on(ocpp_v201_enums.Action.Authorize)
    async def on_authorize(self, **kwargs):
        request = call201.AuthorizePayload(**kwargs)
        logger.debug(f"(Charging Station) {self.id=} <- {request=}")
        response = await self.call(request)
        logger.debug(f"(Charging Station) {self.id=} -> {response=}")
        return response
```

`provisioning_router.py` Authorize handler
```python
@router.on(Action.Authorize)
async def on_authorize(
    *, payload: call.AuthorizePayload, context: HandlerContext
) -> call_result.AuthorizePayload:
    print(context)
    id = context.charging_station_id
    logger.debug(f"(Central System) on_authorize Charging Station {id=}")

    response = call_result.AuthorizePayload(
        id_token_info={"status": AuthorizationStatusType.accepted}
    )   
    logger.debug(f"(Central System) Charging Station {id=} {response=}")
    await context.send(response)
    return response
```

Instead of mocking the response on the CSMS I want to validate the JWT using a basic validation function with a secret key. My problem is that when it comes to send the JWT via URL params (.....&token=<token>) I receive an error that says that the token length is exceeding the maximum allowed. ¿Is there a way to implement/modify this length to allow me to use my webapp JWTs to authorize users? 

Here I'm showing the trace I'm trying to execute:

POST call:
```bash
curl -X 'POST' \
  'http://localhost:8080/authorize/4?token=verylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtoken' \
  -H 'accept: application/json' \
  -d ''
```

Charging station console:
```bash
jsonschema.exceptions.ValidationError: 'verylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtoken' is too long

Failed validating 'maxLength' in schema['properties']['idToken']['properties']['idToken']:
    {'description': 'IdToken is case insensitive. Might hold the hidden id '
                    'of an RFID tag, but can for example also contain a '
                    'UUID.\r\n',
     'maxLength': 36,
     'type': 'string'}

On instance['idToken']['idToken']:
    'verylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtokenverylongJWTtoken'
```

Hope we can solve this problem. Thank you in advance.

**Reply by villekr** [reply]
This question is not anymore related to ocpp-asgi library. So I advice you to study OCPP 2.0.1 specification and semantics of each OCPP request type and what is the direction of request from (from charging station to csms or from csms to charging station).

Authorize request is meant to be sent from charging station to csms. You're doing it here wrong way i.e. other way around.

**Reply by jxtaaa** [reply]
I can't really understand why I'm doing it the wrong way. I think I'm doing Client API -> Charging Station -> CSMS. Isn't that right?

**Reply by villekr** [reply]
No it's not right. You're doing: CSMS --> (Authorize request) --> Charging Station --> (Authorize request) --> CSMS. Client API is part of CSMS if you look at it from Charging Station perspective.

**Reply by jxtaaa** [reply]
I understand what you are saying, then, how can I call from client API directly to CSMS router instead of calling the charging station on ocpp-asgi? Can I write direct code on client_api and take it as part of CSMS as you have already said?

edit: I'm confused, what I need to do is to call directly the charging station ,right? If I have to, how do I call it directly?

**Reply by villekr** [reply]
Client API is in fact part of your CSMS. So you can send any (CSMS to Charging Station type) request to Charging Station directly from your Client API. Router methods are there to handle requests from Charging Station towards CSMS.

**Reply by jxtaaa** [reply]
So can I just validate the token from client api? Would that be right? (Taking client API as part of CSMS, of course)

**Comment by villekr** [comment, marked as answer]
Your JWT token has no meaning in relation to OCPP communication. It's part of authenticating user in Client API. So yes, you probably will need to perform JWT token validation somehow in Client API (and this has nothing to do with OCPP, check e.g. FastAPI docs for instructions how to do that). After authentication you will likely have some form of authorization i.e. whether user is able to perform the specific request towards a specific charging station (once again this user authorization in Client API is something that's not related to OCPP communication as such).

**Reply by jxtaaa** [reply]
That sounds great, I'm gonna work on that. Thank you so much for your help.
